File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :refer [blank? ends-with? trim]]))

(defn shout? [phrase]
  (and
   (not-any? #(Character/isLowerCase %) phrase)
   (some #(Character/isUpperCase %) phrase)))

(defn question? [phrase]
  (ends-with? (trim phrase) "?"))

(defn response-for [phrase]
  (cond (blank? phrase) "Fine. Be that way!"
        (shout? phrase) "Whoa, chill out!"
        (question? phrase) "Sure."
        :else "Whatever."))

File: ecmascript/bob/bob.js
const isUpper = ch => ch.toUpperCase() === ch && ch.toLowerCase() !== ch
const isLower = ch => ch.toLowerCase() === ch && ch.toUpperCase() !== ch
const isShout = message =>
      [...message].some(isUpper) && ![...message].some(isLower);
const isQuestion = message => message.match(/\?$/);
const isSilence = message => message.trim() === "";

export default class Bob {
  hey(message) {
    return isSilence(message) ? 'Fine. Be that way!'
      : isShout(message) ? 'Whoa, chill out!'
      : isQuestion(message) ? 'Sure.'
      : 'Whatever.'
  }
}
